src/main/java/io/nats/client/impl/NatsConnectionWriter.java

```diff
@@ -116,15 +116,13 @@ synchronized void sendMessageBatch(NatsMessage msg, DataPort dataPort, NatsStati
             if (sendPosition + size > sendBuffer.length) {
                 if (sendPosition == 0) { // have to resize
                     this.sendBuffer = new byte[(int)Math.max(sendBuffer.length + size, sendBuffer.length * 2L)];
-                } else { // else send and go to next message
+                } else { // else send and continue with current message
                     dataPort.write(sendBuffer, sendPosition);
                     connection.getNatsStatistics().registerWrite(sendPosition);
                     sendPosition = 0;
-                    msg = msg.next;

```

Comment:
I know it's not obvious, the NatMessage has a field called next, which is really a linked list. You have to get the next message or you would just infinite loop.

Reply:
Yeah exactly. But the thing is that this `msg = msg.next;` statement is already done at the end of the loop.

I believe this line removes the current message, doesn't send it, and just skips to the next message.

Reply:
See also my comment here, which is a more detailed explanation:
https://github.com/nats-io/nats.java/issues/644#issuecomment-1257709076

Reply:
I also commented in 644. I get it now. This is another option, at most the loop will execute twice. Both options are probably equivla
```
while (sendPosition + size > sendBuffer.length) {
    if (sendPosition == 0) { // have to resize
        this.sendBuffer = new byte[(int)Math.max(sendBuffer.length + size, sendBuffer.length * 2L)];
    } else { // else send and go to next message
        dataPort.write(sendBuffer, sendPosition);
        connection.getNatsStatistics().registerWrite(sendPosition);
        sendPosition = 0;
    }
}
```

Reply:
I'm going to ask one other person to check this, make sure I can explain it to them and see if they see it too, but I'm included to accept this change.

Reply:
A unit test to check this exact condition would be great. Not sure how easy it would be, I think a very large message
 and a small `options.getBufferSize()` might do the trick.
 ```
int bufSize = options.getBufferSize();
this.sendBuffer = new byte[bufSize];
```

Reply:
I think so, not quite sure how to test this. As this case would only be triggered if at least one message was sent prior in this batch (so the `sendPosition > 0`). So you'd need one small(er) message that would fit in the buffer to move the `sendPosition`, and then another bigger message that wouldn't fit in the buffer.